dbms/src/Flash/Management/S3Lock.cpp

```diff
@@ -0,0 +1,244 @@
+    req->set_upload_seq(upload_seq);
+
+    auto res = std::make_shared<kvrpcpb::TryAddLockResponse>();
+    auto call = pingcap::kv::RpcCall<kvrpcpb::TryAddLockRequest>(req);
+    auto * cluster = context.getTMTContext().getKVCluster();
+    LOG_DEBUG(log, "Send TryAddLock request, address={} req={}", address, req->DebugString());
+    cluster->rpc_client->sendRequest(address, call, timeout);
+    const auto & resp = call.getResp();
+    LOG_DEBUG(log, "Received TryAddLock response, resp={}", resp->DebugString());
+    if (resp->has_error())
+    {
+        LOG_ERROR(log, "TryMarkDelete get resp with error={}", resp->error().DebugString());
+        return {false, resp->error()};
+    }
+
+    return {resp->is_success(), std::nullopt};
+}
+
+std::pair<bool, std::optional<kvrpcpb::S3LockError>> S3LockClient::sendTryMarkDeleteRequest(String address, int timeout, const String & ori_data_file, UInt32 ori_store_id)
+{
+    auto req = std::make_shared<kvrpcpb::TryMarkDeleteRequest>();
+    req->set_ori_data_file(ori_data_file);
+    req->set_ori_store_id(ori_store_id);
+
+    auto res = std::make_shared<kvrpcpb::TryMarkDeleteResponse>();
+    auto call = pingcap::kv::RpcCall<kvrpcpb::TryMarkDeleteRequest>(req);
+    auto * cluster = context.getTMTContext().getKVCluster();
+    LOG_DEBUG(log, "Send TryMarkDelete request, address={} req={}", address, req->DebugString());
+    cluster->rpc_client->sendRequest(address, call, timeout);
+    const auto & resp = call.getResp();
+    LOG_DEBUG(log, "Received TryMarkDelete response, resp={}", resp->DebugString());
+    if (resp->has_error())
+    {
+        LOG_ERROR(log, "TryMarkDelete get resp with error={}", resp->error().DebugString());
+        return {false, resp->error()};
+    }
+
+    return {resp->is_success(), std::nullopt};
+}
+
+
+bool S3LockService::tryAddLockImpl(const String & ori_data_file, UInt32 ori_store_id, UInt32 lock_store_id, UInt32 upload_seq, kvrpcpb::TryAddLockResponse * response)
+{
+    if (ori_data_file.empty())
+        return false;
+
+    String data_file_name;
+    if (ori_data_file[0] == 't')
+        data_file_name = fmt::format("/s{}/stable/{}", ori_store_id, ori_data_file);
+    else if (ori_data_file[0] == 'd')
+        data_file_name = fmt::format("/s{}/data/{}", ori_store_id, ori_data_file);
+    else
+        return false;
+
+    String delete_file_name = fmt::format("{}.del", data_file_name);
+    String lock_file_name = fmt::format("/s{}/lock/{}.lock_{}_{}", ori_store_id, ori_data_file, lock_store_id, upload_seq);
+
+    // Get the lock of the file
+    std::unordered_map<std::string, DataFileMutexPtr>::iterator it;
+    {
+        std::unique_lock lock(file_latch_map_mutex);
+        it = file_latch_map.find(data_file_name);
+        if (it == file_latch_map.end())
+        {
+            it = file_latch_map.emplace(data_file_name, std::make_shared<DataFileMutex>()).first;
+        }
+    }
+    auto & file_lock = it->second;
```

Comment:
```suggestion
    DataFileMutexPtr file_lock;
    {
        std::unique_lock lock(file_latch_map_mutex);
        it = file_latch_map.find(data_file_name);
        if (it == file_latch_map.end())
        {
            it = file_latch_map.emplace(data_file_name, std::make_shared<DataFileMutex>()).first;
        }
        file_lock = it->second;
        file_lock->addRefCount();
    }
```
Should only return the shared_ptr to upper scope, and should add ref count under the unique_lock of file_latch_map_mutex